src/util/threadpool.rs

```diff
@@ -0,0 +1,618 @@
+        assert!(num_threads >= 1);
+        let (tx, rx) = channel();
+        let rx = Arc::new(Mutex::new(rx));
+        let meta = Arc::new(Mutex::new(ThreadPoolMeta::new(queue, tx)));
+
+        let mut threads = Vec::with_capacity(num_threads);
+        // Threadpool threads
+        for _ in 0..num_threads {
+            threads.push(new_thread(name.clone(), rx.clone(), meta.clone()));
+        }
+
+        ThreadPool {
+            meta: meta.clone(),
+            threads: threads,
+        }
+    }
+
+    /// Executes the function `job` on a thread in the pool.
+    pub fn execute<F>(&mut self, group_id: T, job: F)
+        where F: FnOnce() + Send + 'static
+    {
+        let lock = self.meta.clone();
+        let mut meta = lock.lock().unwrap();
+        meta.push_task(group_id, job);
+    }
+
+    pub fn get_tasks_num(&self) -> usize {
+        let lock = self.meta.clone();
+        let meta = lock.lock().unwrap();
+        meta.get_tasks_num()
+    }
+
+    pub fn stop(&mut self) -> Result<(), String> {
+        {
+            let lock = self.meta.clone();
+            let tasks = lock.lock().unwrap();
+            tasks.push_shutdown_tasks(self.threads.len()).unwrap();
+        }
+        while let Some(t) = self.threads.pop() {
+            if let Err(e) = t.join() {
+                return Err(format!("{:?}", e));
+            }
+        }
+        Ok(())
+    }
+}
+
+// each thread has a worker.
+struct Worker<Q, T> {
+    job_rever: Arc<Mutex<Receiver<bool>>>,
+    pool_meta: Arc<Mutex<ThreadPoolMeta<Q, T>>>,
+}
+
+impl<Q, T> Worker<Q, T>
+    where Q: ScheduleQueue<T>,
+          T: Send + Clone + 'static
+{
+    fn new(receiver: Arc<Mutex<Receiver<bool>>>,
+           pool_meta: Arc<Mutex<ThreadPoolMeta<Q, T>>>)
+           -> Worker<Q, T> {
+        Worker {
+            job_rever: receiver,
+            pool_meta: pool_meta,
+        }
+    }
+
+    // wait to receive info from job_receiver,
+    // return false when get stop msg
+    #[inline]
+    fn wait(&self) -> bool {
+        // try to receive notify
+        let job_receiver = self.job_rever.lock().unwrap();
+        job_receiver.recv().unwrap()
+    }
+
+    #[inline]
+    fn get_next_task(&mut self) -> Option<Task<T>> {
+        // try to get task
+        let mut meta = self.pool_meta.lock().unwrap();
+        meta.pop_next_task()
+    }
+
+    fn run(&mut self) {
+        // start the worker.
+        // loop break on receive stop message.
+        while self.wait() {
+            // handle task
+            // since `tikv` would be down on any panic happens,
+            // we needn't process panic case here.
+            if let Some(task) = self.get_next_task() {
+                let group_id = task.group_id.clone();
+                task.task.call_box(());
+                let mut meta = self.pool_meta.lock().unwrap();
+                meta.finished_task(group_id);
+            }
+        }
+    }
+}
+
+fn new_thread<Q, T>(name: String,
```

Comment:
this function is so simple that I think you can create the thread directly in the new function.